infrastructure/smart-contracts/contracts/upgrades/referee/Referee9.sol

```diff
@@ -421,19 +422,69 @@ contract Referee9 is Initializable, AccessControlEnumerableUpgradeable {
+        // checking the last assertion id here as it will always need to be checked
+        // regardless of if it is a single or batched challenge
         bytes32 comboHash = keccak256(abi.encodePacked(_assertionId, rollupAddress));
         require(!rollupAssertionTracker[comboHash], "9");
         rollupAssertionTracker[comboHash] = true;
+        
+        // Connect to the rollup contract
+        IRollupCore rollup = IRollupCore(rollupAddress);
+
+        // get the node information from the rollup.
+        Node memory currentNode = rollup.getNode(_assertionId);
+        require(currentNode.createdAtBlock == _assertionTimestamp, "12");
+
+        // If the gap is more than 1 assertion, we need to handle as a batch challenge
+        bool isBatch = _assertionId - _predecessorAssertionId > 1;
+
+        if(isBatch){
+
+            // Initialize the array to store the assertionIds
+            uint64 [] memory assertionIds = new uint64[](_assertionId - _predecessorAssertionId);
+
+            // Loop through the assertions and check if they have been submitted
+            // We do not need to check the assertionId as it has already been checked
+            // earlier in the function
+            for(uint64 i = _predecessorAssertionId + 1; i < _assertionId; i++){
+
+                // create the comboHash for the assertionId and rollupAddress
+                bytes32 comboHashBatch = keccak256(abi.encodePacked(i, rollupAddress));
+
+                // check the assertionId and rollupAddress combo haven't been submitted yet
+                require(!rollupAssertionTracker[comboHashBatch], "9");
+
+                // set the comboHash to true to indicate it has been submitted
+                rollupAssertionTracker[comboHashBatch] = true;
 
-        // verify the data inside the hash matched the data pulled from the rollup contract
-        if (isCheckingAssertions) {
+                if (isCheckingAssertions) {
 
-            // get the node information from the rollup.
-            Node memory node = IRollupCore(rollupAddress).getNode(_assertionId);
+                    // get the node information for this assertion from the rollup.
+                    Node memory node = rollup.getNode(i);
 
-            require(node.prevNum == _predecessorAssertionId, "10");
-            require(node.confirmData == _confirmData, "11");
-            require(node.createdAtBlock == _assertionTimestamp, "12");
+                    // check the _predecessorAssertionId is correct
+                    require(node.prevNum == i - 1, "10");   
+                }      
+
+                // add the assertionId to the array
+                assertionIds[i - _predecessorAssertionId - 1] = i;
+            }
+
+            // add the last assertionId to the array sicne it was not included in the loop
+            assertionIds[_assertionId - _predecessorAssertionId - 1] = _assertionId;
```

Comment:
Is there a specific reason we cannot just run the loop for all assertionIds and not have to handle the "last" on its own ?
This makes this code rather hard to read I fear.

Reply:
Refactored to handle all inside of loop!